### DISCUSSION TITLE
Setting request parameter manually in case of content type application/json

### DISCUSSION THREAD

**Discussion by Alireza-Tabatabaeian** [symfony/symfony#48849] [Q&A] [answered]
### Symfony version(s) affected

4.4.2

### Description

Working on a Drupal project I faced an issue, when trying to call an endpoint with content-type set as 'application/json' the endpoint returned invalid arguments error, after some digging I realized that the module contributor (Simple oauth module) has used $request->getParsedBody() (a method of Psr\Http\Message\ServerRequestInterface) to get arguments from request, I tried to pass parameters from json decoded body, but no chance as $request was passing around and keeps making problem.

I had faced this problem in some more modules (like Drupal core Rest module) too, and the reason was the same.

So tracking back I reached to Symfony\Component\HttpFoundation\Request.php and realized that in case of 'application/x-www-form-urlencoded', the required parameters are provided manually, so I thought the same modification can be applied in case of 'application/json'

I added this code in my environment  :

      if (str_starts_with($request->headers->get('CONTENT_TYPE', ''), 'application/json') && in_array(strtoupper($request->server->get('REQUEST_METHOD', 'GET')), ['POST', 'PUT', 'DELETE', 'PATCH'])) {	
            $data = json_decode($request->getContent(),true);
            $request->request = new ParameterBag($data);
        }

and now everything works as expected, but I am not sure about security vulnerability of this hack.  

### How to reproduce

In a drupal project install Simple Oauth module and try to get an access token calling this endpoint:

url: {site url}/oauth/token
method: Post
content-type: application/json
body:
{
"username": "USERNAME",
"password": "PASSWORD",
"client_id": "CLIENT_ID",
"client_secret": "CLIENT_SECRET",
"grant_type": "password"
}

in this case error happens but in case of 'application/x-www-form-urlencoded' or 'application/form-data' endpoint woks with no problem

### Possible Solution

this is the code I added to Symfony\Component\HttpFoundation\Request::createFromGlobals before returning $request

      if (str_starts_with($request->headers->get('CONTENT_TYPE', ''), 'application/json') && in_array(strtoupper($request->server->get('REQUEST_METHOD', 'GET')), ['POST', 'PUT', 'DELETE', 'PATCH'])) {	
            $data = json_decode($request->getContent(),true);
            $request->request = new ParameterBag($data);
        }

**Comment by derrabus** [comment]
I'm moving this to support because we don't maintain the 4.4 branch anymore.

**Reply by Alireza-Tabatabaeian** [reply]
Well actually I checked Symfony 6.2 and this pattern will reproduce there again. in Symfony\Component\HttpFoundation::createFromGlobals line 297 to 302, Symfony manually sets request parameter in case of 'application/x-www-form-urlencoded' request content type, if similar solution applies for 'application/json' too, then there will be less trouble in future steps determining where parameters should be fetched from

**Comment by derrabus** [comment, marked as answer]
A JSON payload and form data are not necessarily the same, so we cannot really assume that the JSON payload contains "request parameters" (whatever that is to you). In some applications, those payloads can get really large, so always decoding them might be a very bad idea.

While you can certainly do that in your application by registering a listener on the request event, I believe we should not provide that functionality in Symfony's core.
